### ISSUE TITLE
P.cpp

### ISSUE THREAD

**Issue by talha01-art** [talha01-art/Real-time-data-processing-#1]
#include <iostream>
#include <fstream>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <optional>
#include <thread>
#include <vector>
#include <functional>
#include <boost/asio.hpp>

using namespace std; // Simplifies code by avoiding std:: prefixes
using boost::asio::ip::tcp;

// Thread-Safe Queue for Data Buffering
template <typename T>
class ThreadSafeQueue {
public:
    void push(T value) {
        lock_guard<mutex> lock(mutex_);
        queue_.push(move(value));
        cond_.notify_one();
    }

    optional<T> pop() {
        unique_lock<mutex> lock(mutex_);
        cond_.wait(lock, [this] { return !queue_.empty() || stop_; });
        if (queue_.empty()) return nullopt;
        T value = move(queue_.front());
        queue_.pop();
        return value;
    }

    void stop() {
        {
            lock_guard<mutex> lock(mutex_);
            stop_ = true;
        }
        cond_.notify_all();
    }

private:
    queue<T> queue_;
    mutex mutex_;
    condition_variable cond_;
    bool stop_ = false;
};

// Thread Pool for Processing Data
class ThreadPool {
public:
    ThreadPool(size_t numThreads, ThreadSafeQueue<string>& queue)
        : queue_(queue) {
        for (size_t i = 0; i < numThreads; ++i) {
            workers_.emplace_back([this] {
                while (auto data = queue_.pop()) {
                    process(*data);
                }
            });
        }
    }

    ~ThreadPool() {
        queue_.stop();
        for (auto& worker : workers_) {
            worker.join();
        }
    }

private:
    void process(const string& data) {
        // Example processing: Convert to uppercase and log
        string processed;
        for (char c : data) {
            processed += toupper(c);
        }
        cout << "[PROCESSED] " << processed << endl;
    }

    ThreadSafeQueue<string>& queue_;
    vector<thread> workers_;
};

// Stream Processing: Read Data from a File
void readFileAsStream(const string& filename, ThreadSafeQueue<string>& queue) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Error: Could not open file " << filename << endl;
        return;
    }

    string line;
    while (getline(file, line)) {
        queue.push(line); // Push data to the queue
        this_thread::sleep_for(chrono::milliseconds(100)); // Simulate real-time delay
    }
    queue.stop(); // Signal end of data
}

// Event-Driven Processing: Network Server
void startServer(ThreadSafeQueue<string>& queue, unsigned short port) {
    boost::asio::io_context io_context;
    tcp::acceptor acceptor(io_context, tcp::endpoint(tcp::v4(), port));

    cout << "Server listening on port " << port << endl;

    while (true) {
        tcp::socket socket(io_context);
        acceptor.accept(socket);

        thread([&, &socket = socket] {
            try {
                char data[1024];
                while (true) {
                    size_t length = socket.read_some(boost::asio::buffer(data));
                    string message(data, length);
                    queue.push(message); // Push incoming message to the queue
                    cout << "[RECEIVED] " << message << endl;
                }
            } catch (exception& e) {
                cerr << "Exception: " << e.what() << endl;
            }
        }).detach();
    }
}

// Main Function
int main() {
    // Create a shared queue for data
    ThreadSafeQueue<string> queue;

    // Start the thread pool (4 worker threads)
    ThreadPool pool(4, queue);

    // Start file stream reader (simulates real-time data)
    thread fileReader(readFileAsStream, "p1.txt", ref(queue)); // Updated filename

    // Start network server (listens on port 12345)
    thread serverThread(startServer, ref(queue), 12345);

    // Wait for threads to finish (in reality, use a graceful shutdown)
    fileReader.join();
    serverThread.join();

    return 0;
}
